File: ios/ICEInfo/Network/StationFacilitiesRepository.swift
import Foundation

actor StationFacilitiesRepository {
    static let shared = StationFacilitiesRepository()

    private let stadaBase = "https://apis.deutschebahn.com/db-api-marketplace/apis/station-data/v2"
    private let fastaBase = "https://apis.deutschebahn.com/db-api-marketplace/apis/fasta/v2"

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    func searchStations(query: String) async -> [StationSearchResult] {
        guard query.count >= 4 else { return [] }
        guard let url = URL(string: "\(stadaBase)/stations?searchstring=\(query)*&limit=10"
            .addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "") else { return [] }
        do {
            var request = URLRequest(url: url, timeoutInterval: 10)
            request.setValue(BuildConfig.dbClientId, forHTTPHeaderField: "DB-Client-ID")
            request.setValue(BuildConfig.dbApiKey, forHTTPHeaderField: "DB-Api-Key")
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return [] }
            let stada = try decoder.decode(StadaResponse.self, from: data)
            return stada.result.compactMap { station in
                guard let name = station.name, let eva = station.evaNumbers.first?.number else { return nil }
                return StationSearchResult(evaNr: "\(eva)", name: name)
            }
        } catch {
            return []
        }
    }

    func fetchFacilities(evaNr: String, stationName: String) async -> StationInfo {
        do {
            async let stadaResult = fetchStada(evaNr: evaNr)
            let result = await stadaResult
            let liveFacilities: [StationFacility]
            if let stationNumber = result?.stationNumber {
                liveFacilities = await fetchFasta(stationNumber: stationNumber)
            } else {
                liveFacilities = []
            }
            return StationInfo(
                evaNr: evaNr,
                name: stationName,
                liveFacilities: liveFacilities,
                staticFacilities: result?.toStaticFacilityTypes() ?? [],
                isLoading: false
            )
        }
    }

    private func fetchStada(evaNr: String) async -> StadaStation? {
        guard let url = URL(string: "\(stadaBase)/stations?eva=\(evaNr)&limit=1") else { return nil }
        do {
            var request = URLRequest(url: url, timeoutInterval: 10)
            request.setValue(BuildConfig.dbClientId, forHTTPHeaderField: "DB-Client-ID")
            request.setValue(BuildConfig.dbApiKey, forHTTPHeaderField: "DB-Api-Key")
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            let stada = try decoder.decode(StadaResponse.self, from: data)
            return stada.result.first
        } catch {
            return nil
        }
    }

    private func fetchFasta(stationNumber: Int) async -> [StationFacility] {
        guard let url = URL(string: "\(fastaBase)/facilities?stationnumber=\(stationNumber)&type=ELEVATOR,ESCALATOR") else { return [] }
        do {
            var request = URLRequest(url: url, timeoutInterval: 10)
            request.setValue(BuildConfig.dbClientId, forHTTPHeaderField: "DB-Client-ID")
            request.setValue(BuildConfig.dbApiKey, forHTTPHeaderField: "DB-Api-Key")
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return [] }
            let facilities = try decoder.decode([FastaFacility].self, from: data)
            return facilities.map { fac in
                StationFacility(
                    id: "\(fac.equipmentnumber)",
                    type: (fac.type?.uppercased() == "ESCALATOR") ? .escalator : .elevator,
                    label: fac.description ?? fac.type?.lowercased().prefix(1).uppercased() ?? "Anlage",
                    status: fac.state?.uppercased() == "ACTIVE" ? .active :
                            fac.state?.uppercased() == "INACTIVE" ? .inactive : .unknown,
                    description: fac.stateExplanation ?? ""
                )
            }
        } catch {
            return []
        }
    }
}

private extension StadaStation {
    func toStaticFacilityTypes() -> [FacilityType] {
        var types: [FacilityType] = []
        if hasPublicFacilities == true { types.append(.toilet) }
        if hasWiFi == true { types.append(.wifi) }
        if let svc = hasMobilityService, svc != "no" { types.append(.infoDesk) }
        if hasLockerSystem == true { types.append(.waitingRoom) }
        if hasBicycleParking == true { types.append(.bikeParking) }
        if hasParking == true { types.append(.parking) }
        return types
    }
}

File: app/src/main/java/com/nruge/iceinfo/StationFacilitiesRepository.kt
package com.nruge.iceinfo

import android.util.Log
import com.nruge.iceinfo.model.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

object StationFacilitiesRepository {

    private const val STADA_BASE  = "https://apis.deutschebahn.com/db-api-marketplace/apis/station-data/v2"
    private const val FASTA_BASE  = "https://apis.deutschebahn.com/db-api-marketplace/apis/fasta/v2"

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) {
            json(json, contentType = ContentType.Any)
        }
        install(HttpTimeout) {
            requestTimeoutMillis = 10_000
            connectTimeoutMillis = 5_000
            socketTimeoutMillis = 10_000
        }
        defaultRequest {
            header("Accept", "application/json")
        }
    }

    // ── Station search via StaDa ─────────────────────────────────────────────

    suspend fun searchStations(query: String): List<StationSearchResult> = withContext(Dispatchers.IO) {
        if (query.length < 4) return@withContext emptyList()
        try {
            val httpResponse = client.get("$STADA_BASE/stations") {
                header("DB-Client-ID", BuildConfig.DB_CLIENT_ID)
                header("DB-Api-Key", BuildConfig.DB_CLIENT_SECRET)
                parameter("searchstring", "$query*")
                parameter("limit", 10)
            }
            if (!httpResponse.status.isSuccess()) return@withContext emptyList()
            val response: StadaResponse = httpResponse.body()
            response.result.mapNotNull { station ->
                val name = station.name ?: return@mapNotNull null
                val eva  = station.evaNumbers.firstOrNull()?.number ?: return@mapNotNull null
                StationSearchResult(evaNr = eva.toString(), name = name)
            }
        } catch (e: Exception) {
            Log.e("StationFacilities", "searchStations failed: ${e::class.simpleName} — ${e.message}")
            emptyList()
        }
    }

    // ── Facilities: StaDa (static) + FaSta (live) ───────────────────────────

    suspend fun fetchFacilities(evaNr: String, stationName: String): StationInfo = withContext(Dispatchers.IO) {
        try {
            // 1. StaDa: EVA → DB station number + static features
            val stadaResult = fetchStada(evaNr)

            // 2. FaSta: live elevator/escalator status (needs DB station number)
            val liveFacilities = stadaResult?.stationNumber
                ?.let { fetchFasta(it) }
                ?: emptyList()

            StationInfo(
                evaNr = evaNr,
                name = stationName,
                liveFacilities = liveFacilities,
                staticFacilities = stadaResult?.toStaticFacilityTypes() ?: emptyList()
            )
        } catch (e: Exception) {
            Log.e("StationFacilities", "fetchFacilities failed: ${e::class.simpleName} — ${e.message}")
            StationInfo(evaNr = evaNr, name = stationName, error = "Daten konnten nicht geladen werden.")
        }
    }

    // ── StaDa ────────────────────────────────────────────────────────────────

    private suspend fun fetchStada(evaNr: String): StadaStation? {
        return try {
            val httpResponse = client.get("$STADA_BASE/stations") {
                header("DB-Client-ID", BuildConfig.DB_CLIENT_ID)
                header("DB-Api-Key", BuildConfig.DB_CLIENT_SECRET)
                parameter("eva", evaNr.toLongOrNull() ?: evaNr)
                parameter("limit", 1)
            }
            if (!httpResponse.status.isSuccess()) {
                Log.e("StationFacilities", "StaDa HTTP ${httpResponse.status} for EVA $evaNr — ${httpResponse.bodyAsText()}")
                return null
            }
            val response: StadaResponse = httpResponse.body()
            Log.d("StationFacilities", "StaDa returned ${response.result.size} station(s) for EVA $evaNr")
            response.result.firstOrNull()
        } catch (e: Exception) {
            Log.e("StationFacilities", "StaDa exception for EVA $evaNr: ${e::class.simpleName} — ${e.message}")
            null
        }
    }

    private fun StadaStation.toStaticFacilityTypes(): List<FacilityType> = buildList {
        if (hasPublicFacilities == true)                    add(FacilityType.TOILET)
        if (hasWiFi == true)                               add(FacilityType.WIFI)
        if (hasMobilityService != null
            && hasMobilityService != "no")                 add(FacilityType.INFO_DESK)
        if (hasLockerSystem == true)                       add(FacilityType.WAITING_ROOM)
        if (hasBicycleParking == true)                     add(FacilityType.BIKE_PARKING)
        if (hasParking == true)                            add(FacilityType.PARKING)
    }

    // ── FaSta ────────────────────────────────────────────────────────────────

    private suspend fun fetchFasta(stationNumber: Int): List<StationFacility> {
        return try {
            val httpResponse = client.get("$FASTA_BASE/facilities") {
                header("DB-Client-ID", BuildConfig.DB_CLIENT_ID)
                header("DB-Api-Key", BuildConfig.DB_CLIENT_SECRET)
                parameter("stationnumber", stationNumber)
                parameter("type", "ELEVATOR,ESCALATOR")
            }
            if (!httpResponse.status.isSuccess()) {
                Log.e("StationFacilities", "FaSta HTTP ${httpResponse.status} for station $stationNumber — ${httpResponse.bodyAsText()}")
                return emptyList()
            }
            val facilities: List<FastaFacility> = httpResponse.body()
            Log.d("StationFacilities", "FaSta returned ${facilities.size} facilities for station $stationNumber")
            facilities.map { it.toStationFacility() }
        } catch (e: Exception) {
            Log.e("StationFacilities", "FaSta exception for station $stationNumber: ${e::class.simpleName} — ${e.message}")
            emptyList()
        }
    }

    private fun FastaFacility.toStationFacility() = StationFacility(
        id = equipmentNumber.toString(),
        type = when (type?.uppercase()) {
            "ELEVATOR"  -> FacilityType.ELEVATOR
            "ESCALATOR" -> FacilityType.ESCALATOR
            else        -> FacilityType.ELEVATOR
        },
        label = description.orEmpty().ifBlank { type?.lowercase()?.replaceFirstChar { it.uppercase() } ?: "Anlage" },
        status = when (state?.uppercase()) {
            "ACTIVE"   -> FacilityStatus.ACTIVE
            "INACTIVE" -> FacilityStatus.INACTIVE
            else       -> FacilityStatus.UNKNOWN
        },
        description = stateExplanation.orEmpty()
    )

    // ── Response models ──────────────────────────────────────────────────────

    @Serializable
    private data class StadaResponse(
        val result: List<StadaStation> = emptyList(),
        val total: Int = 0
    )

    @Serializable
    private data class StadaStation(
        val number: Int? = null,
        val name: String? = null,
        val evaNumbers: List<EvaNumber> = emptyList(),
        val hasWiFi: Boolean? = null,
        val hasPublicFacilities: Boolean? = null,
        val hasBicycleParking: Boolean? = null,
        val hasLockerSystem: Boolean? = null,
        val hasMobilityService: String? = null,
        val hasParking: Boolean? = null,
        val hasLostAndFound: Boolean? = null,
    ) {
        val stationNumber: Int? get() = number
    }

    @Serializable
    private data class EvaNumber(
        val number: Long? = null,
        val isMain: Boolean? = null
    )

    @Serializable
    private data class FastaFacility(
        @SerialName("equipmentnumber") val equipmentNumber: Long = 0L,
        val type: String? = null,
        val state: String? = null,
        val stateExplanation: String? = null,
        val description: String? = null,
        @SerialName("stationnumber") val stationNumber: Int? = null,
    )
}
